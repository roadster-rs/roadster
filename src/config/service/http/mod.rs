#[mockall_double::double]
use crate::app_context::AppContext;
use crate::config::service::http::initializer::Initializer;
use crate::config::service::http::middleware::Middleware;
use crate::util::serde_util::default_true;
use serde_derive::{Deserialize, Serialize};
use validator::{Validate, ValidationError};

pub mod initializer;
pub mod middleware;

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct HttpServiceConfig {
    #[serde(flatten)]
    #[validate(nested)]
    pub address: Address,
    #[serde(default)]
    #[validate(nested)]
    pub middleware: Middleware,
    #[serde(default)]
    #[validate(nested)]
    pub initializer: Initializer,
    #[serde(default)]
    #[validate(nested)]
    pub default_routes: DefaultRoutes,
}

#[derive(Debug, Clone, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct Address {
    pub host: String,
    pub port: u32,
}

impl Address {
    pub fn url(&self) -> String {
        format!("{}:{}", self.host, self.port)
    }
}

#[derive(Debug, Clone, Default, Validate, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
#[validate(schema(function = "validate_default_routes"))]
pub struct DefaultRoutes {
    #[serde(default = "default_true")]
    pub default_enable: bool,
    #[serde(default)]
    pub ping: DefaultRouteConfig,
    #[serde(default)]
    pub health: DefaultRouteConfig,
    #[cfg(feature = "open-api")]
    #[serde(default)]
    pub api_schema: DefaultRouteConfig,
    #[cfg(feature = "open-api")]
    #[serde(default)]
    pub scalar: DefaultRouteConfig,
    #[cfg(feature = "open-api")]
    #[serde(default)]
    pub redoc: DefaultRouteConfig,
}

fn validate_default_routes(
    // This parameter isn't used for some feature flag combinations
    #[allow(unused)] default_routes: &DefaultRoutes,
) -> Result<(), ValidationError> {
    #[cfg(feature = "open-api")]
    {
        let default_enable = default_routes.default_enable;
        let api_schema_enabled = default_routes.api_schema.enable.unwrap_or(default_enable);
        let scalar_enabled = default_routes.scalar.enable.unwrap_or(default_enable);
        let redoc_enabled = default_routes.redoc.enable.unwrap_or(default_enable);

        if scalar_enabled && !api_schema_enabled {
            return Err(ValidationError::new(
                "The Open API schema route must be enabled in order to use the Scalar docs route.",
            ));
        }
        if redoc_enabled && !api_schema_enabled {
            return Err(ValidationError::new(
                "The Open API schema route must be enabled in order to use the Redoc docs route.",
            ));
        }
    }

    Ok(())
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct DefaultRouteConfig {
    pub enable: Option<bool>,
    pub route: Option<String>,
}

impl DefaultRouteConfig {
    pub fn enabled<S>(&self, context: &AppContext<S>) -> bool {
        self.enable.unwrap_or(
            context
                .config()
                .service
                .http
                .custom
                .default_routes
                .default_enable,
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case(false, false)]
    #[case(true, false)]
    #[cfg(not(feature = "open-api"))]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn validate_default_routes(#[case] default_enable: bool, #[case] validation_error: bool) {
        // Arrange
        #[allow(clippy::field_reassign_with_default)]
        let config = {
            let mut config = DefaultRoutes::default();
            config.default_enable = default_enable;
            config
        };

        // Act
        let result = config.validate();

        // Assert
        assert_eq!(result.is_err(), validation_error);
    }

    #[rstest]
    #[case(false, None, None, None, false)]
    #[case(true, None, None, None, false)]
    #[case(false, None, Some(true), None, true)]
    #[case(false, None, None, Some(true), true)]
    #[case(false, Some(true), Some(true), None, false)]
    #[case(false, Some(true), None, Some(true), false)]
    #[cfg(feature = "open-api")]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn validate_default_routes(
        #[case] default_enable: bool,
        #[case] api_schema_enabled: Option<bool>,
        #[case] scalar_enabled: Option<bool>,
        #[case] redoc_enabled: Option<bool>,
        #[case] validation_error: bool,
    ) {
        // Arrange
        #[allow(clippy::field_reassign_with_default)]
        let config = {
            let mut config = DefaultRoutes::default();
            config.default_enable = default_enable;
            config.api_schema.enable = api_schema_enabled;
            config.scalar.enable = scalar_enabled;
            config.redoc.enable = redoc_enabled;
            config
        };

        // Act
        let result = config.validate();

        // Assert
        assert_eq!(result.is_err(), validation_error);
    }
}
