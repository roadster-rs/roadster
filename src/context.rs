use async_trait::async_trait;

use crate::config::app_config::AppConfig;

#[derive(Debug, Clone)]
pub struct AppContext {
    pub config: AppConfig,
}

impl AppContext {
    pub async fn new(config: AppConfig) -> anyhow::Result<Self> {
        let context = Self { config };
        Ok(context)
    }
}

// Define a custom trait because the standard From/Into traits don't support async
#[async_trait]
pub trait FromAppContextRef {
    async fn from_app_context_ref(ctx: &AppContext) -> anyhow::Result<Box<Self>>;
}

// Define a custom trait because the standard From/Into traits don't support async
#[async_trait]
pub trait GetAppContextFromRef {
    async fn get_app_context(&self) -> anyhow::Result<AppContext>;
}

// Implement for `AppContext` so consumers can just use `AppContext` for their state if they want
#[async_trait]
impl FromAppContextRef for AppContext {
    async fn from_app_context_ref(ctx: &AppContext) -> anyhow::Result<Box<Self>> {
        Ok(Box::new(ctx.clone()))
    }
}

// Implement for `AppContext` so consumers can just use `AppContext` for their state if they want
#[async_trait]
impl GetAppContextFromRef for AppContext {
    async fn get_app_context(&self) -> anyhow::Result<AppContext> {
        Ok(self.clone())
    }
}
