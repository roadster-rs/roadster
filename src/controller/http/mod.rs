use crate::app_context::AppContext;
#[cfg(feature = "open-api")]
use aide::axum::ApiRouter;
use axum::Router;
use itertools::Itertools;

#[cfg(feature = "open-api")]
pub mod docs;
pub mod health;
pub mod ping;

pub fn build_path(parent: &str, child: &str) -> String {
    // Clean the path to make sure it is valid:
    // 1. Remove any occurrences of double `/`, e.g. `/foo//bar`
    // 2. Remove any trailing `/`
    // 3. Make sure the path starts with a `/`
    let path = format!("{parent}/{child}");
    let path = path.split('/').filter(|s| !s.is_empty()).join("/");
    let path = format!("/{path}");
    path
}

pub fn default_routes<S>(parent: &str, context: &AppContext<S>) -> Router<AppContext<S>>
where
    S: Clone + Send + Sync + 'static,
{
    Router::new()
        .merge(ping::routes(parent, context))
        .merge(health::routes(parent, context))
}

#[cfg(feature = "open-api")]
pub fn default_api_routes<S>(parent: &str, context: &AppContext<S>) -> ApiRouter<AppContext<S>>
where
    S: Clone + Send + Sync + 'static,
{
    ApiRouter::new()
        .merge(ping::api_routes(parent, context))
        .merge(health::api_routes(parent, context))
        // The docs route is only available when using Aide
        .merge(docs::routes(parent, context))
}
