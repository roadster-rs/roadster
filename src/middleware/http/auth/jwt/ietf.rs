use crate::middleware::http::auth::jwt::Subject;
use crate::util::serde::UriOrString;
use chrono::serde::ts_seconds;
use chrono::{DateTime, Utc};
use serde_derive::{Deserialize, Serialize};
use serde_json::Value;
use serde_with::serde_as;
use std::collections::BTreeMap;
use typed_builder::TypedBuilder;

/// JWT Claims. Provides fields for the default/recommended registered claim names. Additional
/// claim names are collected in the `custom` map.
/// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4>
#[serde_as]
#[derive(Debug, Clone, Deserialize, Serialize, TypedBuilder)]
#[non_exhaustive]
pub struct Claims {
    /// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.1>
    #[serde(rename = "iss")]
    #[builder(default, setter(strip_option))]
    pub issuer: Option<UriOrString>,

    /// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.2>
    #[serde(rename = "sub")]
    #[builder(default, setter(strip_option))]
    pub subject: Option<Subject>,

    /// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.3>
    #[serde(rename = "aud", default, skip_serializing_if = "Vec::is_empty")]
    #[serde_as(deserialize_as = "serde_with::OneOrMany<_>")]
    #[builder(default)]
    pub audience: Vec<UriOrString>,

    /// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.4>
    /// Not technically required by the JWT spec, but is required by the default [jsonwebtoken::Validation] we use.
    #[serde(rename = "exp", with = "ts_seconds")]
    pub expires_at: DateTime<Utc>,

    /// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.5>
    #[serde(rename = "nbf")]
    #[serde_as(as = "Option<serde_with::TimestampSeconds>")]
    #[builder(default, setter(strip_option))]
    pub not_before: Option<DateTime<Utc>>,

    /// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.6>
    #[serde(rename = "iat")]
    #[serde_as(as = "Option<serde_with::TimestampSeconds>")]
    #[builder(default, setter(strip_option))]
    pub issued_at: Option<DateTime<Utc>>,

    /// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.7>
    #[serde(rename = "jti")]
    #[builder(default, setter(strip_option))]
    pub jwt_id: Option<String>,

    #[serde(flatten)]
    #[builder(default)]
    pub custom: BTreeMap<String, Value>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::error::RoadsterResult;
    use crate::middleware::http::auth::jwt::decode_auth_token;
    use crate::util::serde::{UriOrString, Wrapper};
    use chrono::{TimeDelta, Utc};
    use jsonwebtoken::{encode, EncodingKey, Header, TokenData};
    use serde_json::from_str;
    use std::ops::{Add, Sub};
    use std::str::FromStr;
    use url::Url;

    const TEST_JWT_SECRET: &str = "test-jwt-secret";
    const AUDIENCE: &[&str] = &["authenticated"];
    const REQUIRED_CLAIMS: &[&str] = &[];

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn decode_token() {
        let jwt = build_token(false, None);

        let decoded: TokenData<Claims> =
            decode_auth_token(&jwt.1, TEST_JWT_SECRET, AUDIENCE, REQUIRED_CLAIMS).unwrap();

        assert_eq!(decoded.claims.subject, jwt.0.subject);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn decode_token_expired() {
        let (_, jwt) = build_token(true, None);

        let decoded: RoadsterResult<TokenData<Claims>> =
            decode_auth_token(&jwt, TEST_JWT_SECRET, AUDIENCE, REQUIRED_CLAIMS);

        assert!(decoded.is_err());
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn decode_token_wrong_audience() {
        let (_, jwt) = build_token(false, Some("different-audience".to_string()));

        let decoded: RoadsterResult<TokenData<Claims>> =
            decode_auth_token(&jwt, TEST_JWT_SECRET, AUDIENCE, REQUIRED_CLAIMS);

        assert!(decoded.is_err());
    }

    fn build_token(expired: bool, audience: Option<String>) -> (Claims, String) {
        let (expires_at, issued_at) = if expired {
            (
                Utc::now().sub(TimeDelta::try_minutes(30).unwrap()),
                Utc::now().sub(TimeDelta::try_minutes(2).unwrap()),
            )
        } else {
            (Utc::now().add(TimeDelta::try_hours(1).unwrap()), Utc::now())
        };
        let claims = Claims {
            issuer: Some(UriOrString::Uri(
                Url::from_str("https://example.com").unwrap(),
            )),
            subject: Some(Subject::Uuid(uuid::Uuid::new_v4())),
            audience: vec![UriOrString::String(
                audience.unwrap_or_else(|| "authenticated".to_string()),
            )],
            expires_at,
            issued_at: Some(issued_at),
            not_before: None,
            jwt_id: None,
            custom: Default::default(),
        };
        let token = encode(
            &Header::default(),
            &claims,
            &EncodingKey::from_secret(TEST_JWT_SECRET.as_ref()),
        )
        .unwrap();
        (claims, token)
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_audience_as_vec() {
        let value: Wrapper<Vec<UriOrString>> =
            from_str(r#"{"inner": ["https://example.com", "aud2"]}"#).unwrap();
        assert_eq!(
            value.inner,
            vec![
                UriOrString::Uri(Url::from_str("https://example.com").unwrap()),
                UriOrString::String("aud2".to_string())
            ]
        );
    }
}
