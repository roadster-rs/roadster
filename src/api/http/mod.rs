use crate::app::context::AppContext;
#[cfg(feature = "open-api")]
use aide::axum::ApiRouter;
use axum::Router;
use axum_core::extract::FromRef;
use itertools::Itertools;

#[cfg(feature = "open-api")]
pub mod docs;
pub mod health;
pub mod ping;

pub fn build_path(parent: &str, child: &str) -> String {
    // Clean the path to make sure it is valid:
    // 1. Remove any occurrences of double `/`, e.g. `/foo//bar`
    // 2. Remove any trailing `/`
    // 3. Make sure the path starts with a `/`
    let path = format!("{parent}/{child}");
    let path = path.split('/').filter(|s| !s.is_empty()).join("/");
    let path = format!("/{path}");
    path
}

pub fn default_routes<S>(parent: &str, state: &S) -> Router<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    Router::new()
        .merge(ping::routes(parent, state))
        .merge(health::routes(parent, state))
}

#[cfg(feature = "open-api")]
pub fn default_api_routes<S>(parent: &str, state: &S) -> ApiRouter<S>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
{
    ApiRouter::new()
        .merge(ping::api_routes(parent, state))
        .merge(health::api_routes(parent, state))
        // The docs route is only available when using Aide
        .merge(docs::routes(parent, state))
}
