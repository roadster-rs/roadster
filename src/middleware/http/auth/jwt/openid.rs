use crate::middleware::http::auth::jwt::Subject;
use chrono::serde::ts_seconds;
use chrono::{DateTime, Utc};
use serde_derive::{Deserialize, Serialize};
use serde_json::Value;
use serde_with::serde_as;
use std::collections::BTreeMap;
use url::Url;

use crate::util::serde_util::{deserialize_from_str, serialize_to_str, UriOrString};

/// JWT Claims. Provides fields for the default/recommended registered claim names. Additional
/// claim names are collected in the `custom` map.
/// See: <https://openid.net/specs/openid-connect-core-1_0.html#IDToken>
#[serde_as]
#[derive(Debug, Deserialize, Serialize)]
pub struct Claims {
    #[serde(rename = "iss")]
    pub issuer: Url,

    #[serde(rename = "sub")]
    pub subject: Subject,

    #[serde(rename = "aud", default, skip_serializing_if = "Vec::is_empty")]
    #[serde_as(deserialize_as = "serde_with::OneOrMany<_>")]
    pub audience: Vec<UriOrString>,

    #[serde(rename = "exp", with = "ts_seconds")]
    pub expires_at: DateTime<Utc>,

    #[serde(rename = "iat", with = "ts_seconds")]
    pub issued_at: DateTime<Utc>,

    #[serde_as(as = "Option<serde_with::TimestampSeconds>")]
    pub auth_time: Option<DateTime<Utc>>,

    pub nonce: Option<String>,

    #[serde(rename = "acr")]
    pub auth_cxt_class_reference: Option<Acr>,

    #[serde(rename = "amr", default, skip_serializing_if = "Vec::is_empty")]
    pub auth_methods_references: Vec<String>,

    /// Note per the OpenID docs: "\[...\] in practice, the azp Claim only occurs when extensions
    /// beyond the scope of this specification are used; therefore, implementations not using such
    /// extensions are encouraged to not use azp and to ignore it when it does occur."
    #[serde(rename = "azp")]
    pub authorized_party: Option<UriOrString>,

    #[serde(flatten)]
    pub custom: BTreeMap<String, Value>,
}

#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(untagged)]
pub enum Acr {
    Uri(Url),
    Int(
        #[serde(
            deserialize_with = "deserialize_from_str",
            serialize_with = "serialize_to_str"
        )]
        u64,
    ),
    String(String),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::util::serde_util::Wrapper;
    use serde_json::from_str;
    use std::str::FromStr;
    use url::Url;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_acr_as_uri() {
        let value: Wrapper<Acr> = from_str(r#"{"inner": "https://example.com"}"#).unwrap();
        assert_eq!(
            value.inner,
            Acr::Uri(Url::from_str("https://example.com").unwrap())
        );
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_acr_as_int() {
        let num = 100;
        let value: Wrapper<Acr> = from_str(&format!(r#"{{"inner": "{num}"}}"#)).unwrap();
        assert_eq!(value.inner, Acr::Int(num));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn serialize_arc_int_as_string() {
        let num = 100;
        let value = Wrapper {
            inner: Acr::Int(num),
        };
        let s = serde_json::to_string(&value).unwrap();
        assert_eq!(s, format!(r#"{{"inner":"{num}"}}"#));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_acr_as_string() {
        let value: Wrapper<Acr> = from_str(r#"{"inner": "invalid-uri"}"#).unwrap();
        assert_eq!(value.inner, Acr::String("invalid-uri".to_string()));
    }
}
