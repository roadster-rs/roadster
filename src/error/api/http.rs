use crate::error::Error;
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Json;
use serde_derive::{Deserialize, Serialize};
use std::fmt::{Display, Formatter};

/// Error type representing an HTTP API error. This is generally expected to be returned explicitly
/// by your application logic.
///
/// # Examples
///
/// ## Alternative -- directly use `StatusCode`
/// If you simply need to create an Axum response with just a status code, this class is not
/// necessary. You can instead use `StatusCode` directly:
///
/// ```rust
/// use axum::http::StatusCode;
/// use axum::response::IntoResponse;
///
/// fn api_method() -> impl IntoResponse {
///     StatusCode::BAD_REQUEST
/// }
/// ```
///
/// This can also work when your api method returns a result, either with a generic response:
///
/// ```rust
/// use axum::http::StatusCode;
/// use axum::response::IntoResponse;
///
/// fn api_method() -> Result<(), impl IntoResponse> {
///     Err(StatusCode::BAD_REQUEST)
/// }
/// ```
///
/// Or when returning a [roadster result][crate::error::RoadsterResult] (which uses a
/// [roadster error][enum@Error] for its `Error` type).
///
/// ```rust
/// use axum::http::StatusCode;
/// use axum::response::IntoResponse;
/// use roadster::error::RoadsterResult;
///
/// fn api_method() -> RoadsterResult<()> {
///     Err(StatusCode::BAD_REQUEST.into())
/// }
/// ```
///
/// ## Create from `StatusCode`
///
/// ```rust
/// use axum::http::StatusCode;
/// use roadster::error::api::http::HttpError;
///
/// let err: HttpError = StatusCode::BAD_REQUEST.into();
/// ```
///
/// ## Create using a helper method
///
/// ```rust
/// use roadster::error::api::http::HttpError;
///
/// let err = HttpError::bad_request();
/// ```
///
/// ## Populate additional fields with builder-style methods
///
/// ```rust
/// use roadster::error::api::http::HttpError;
///
/// let err = HttpError::bad_request()
///     .error("Something went wrong")
///     .details("Field 'A' is missing");
/// ```
///
/// ## Using in an API method
///
/// ```rust
/// use axum::response::IntoResponse;
/// use roadster::error::api::http::HttpError;
///
/// fn api_method() -> Result<(), impl IntoResponse> {
///     let err = HttpError::bad_request()
///         .error("Something went wrong")
///         .details("Field 'A' is missing");
///     Err(err)
/// }
/// ```
///
/// ## Using in an API method that returns `RoadsterResult`
///
/// ```rust
/// use axum::response::IntoResponse;
/// use roadster::error::api::http::HttpError;
/// use roadster::error::RoadsterResult;
///
/// fn api_method() -> RoadsterResult<()> {
///     let err = HttpError::bad_request()
///         .error("Something went wrong")
///         .details("Field 'A' is missing");
///     Err(err.into())
/// }
/// ```
///
#[derive(Debug, Error, Serialize, Deserialize)]
#[cfg_attr(feature = "open-api", derive(aide::OperationIo, schemars::JsonSchema))]
#[non_exhaustive]
pub struct HttpError {
    /// The HTTP status code for the error.
    ///
    /// When this error is converted to an HTTP response, this field is set as the HTTP response
    /// status header and omitted from the response body/payload.
    #[serde(skip)]
    pub status: StatusCode,
    /// Basic description of the error that occurred.
    // Todo: auto-redact sensitive data
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    /// Additional details for the error.
    // Todo: auto-redact sensitive data
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<String>,
    /// The original error. This can be logged to help with debugging, but it is omitted
    /// from the response body/payload to avoid exposing sensitive details from the stacktrace
    /// to the user.
    #[source]
    #[serde(skip)]
    pub source: Option<Box<dyn std::error::Error + Send + Sync>>,
}

impl Display for HttpError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Http Error {} - {:?}", self.status, self.error)
    }
}

impl HttpError {
    pub fn new(status: StatusCode) -> Self {
        Self {
            status,
            error: None,
            details: None,
            source: None,
        }
    }

    /// Utility method to convert this [HttpError] into an [enum@Error].
    pub fn to_err(self) -> Error {
        self.into()
    }

    pub fn error(self, error: impl ToString) -> Self {
        Self {
            error: Some(error.to_string()),
            ..self
        }
    }

    pub fn details(self, details: impl ToString) -> Self {
        Self {
            details: Some(details.to_string()),
            ..self
        }
    }

    pub fn source(self, source: impl std::error::Error + Send + Sync + 'static) -> Self {
        Self {
            source: Some(Box::new(source)),
            ..self
        }
    }

    // Common 4xx errors

    /// Helper method to create an error with status code [StatusCode::BAD_REQUEST]
    pub fn bad_request() -> Self {
        Self::new(StatusCode::BAD_REQUEST)
    }

    /// Helper method to create an error with status code [StatusCode::UNAUTHORIZED]
    pub fn unauthorized() -> Self {
        Self::new(StatusCode::UNAUTHORIZED)
    }

    /// Helper method to create an error with status code [StatusCode::FORBIDDEN]
    pub fn forbidden() -> Self {
        Self::new(StatusCode::FORBIDDEN)
    }

    /// Helper method to create an error with status code [StatusCode::NOT_FOUND]
    pub fn not_found() -> Self {
        Self::new(StatusCode::NOT_FOUND)
    }

    /// Helper method to create an error with status code [StatusCode::GONE]
    pub fn gone() -> Self {
        Self::new(StatusCode::GONE)
    }

    // Common 5xx errors

    /// Helper method to create an error with status code [StatusCode::INTERNAL_SERVER_ERROR]
    pub fn internal_server_error() -> Self {
        Self::new(StatusCode::INTERNAL_SERVER_ERROR)
    }
}

impl From<StatusCode> for HttpError {
    fn from(value: StatusCode) -> Self {
        HttpError::new(value)
    }
}

impl From<StatusCode> for Error {
    fn from(value: StatusCode) -> Self {
        HttpError::new(value).into()
    }
}

impl IntoResponse for HttpError {
    fn into_response(self) -> Response {
        let status = self.status;
        let mut res = Json(self).into_response();
        *res.status_mut() = status;
        res
    }
}
