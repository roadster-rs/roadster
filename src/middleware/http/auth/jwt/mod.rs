#[cfg(feature = "jwt-ietf")]
pub mod ietf;
#[cfg(feature = "jwt-openid")]
pub mod openid;

use crate::app::context::AppContext;
use crate::error::api::http::HttpError;
use crate::error::{Error, RoadsterResult};
#[cfg(feature = "jwt-ietf")]
use crate::middleware::http::auth::jwt::ietf::Claims;
#[cfg(all(feature = "jwt-openid", not(feature = "jwt-ietf")))]
use crate::middleware::http::auth::jwt::openid::Claims;
use crate::util::serde::{deserialize_from_str, serialize_to_str};
use async_trait::async_trait;
use axum::extract::{FromRef, FromRequestParts};
use axum::http::header::AUTHORIZATION;
use axum::http::request::Parts;
use axum::RequestPartsExt;
use axum_extra::extract::CookieJar;
use axum_extra::headers::authorization::Bearer;
use axum_extra::headers::{Authorization, HeaderValue};
use axum_extra::TypedHeader;
use itertools::Itertools;
use jsonwebtoken::{decode, DecodingKey, Header, TokenData, Validation};
use serde_derive::{Deserialize, Serialize};
#[cfg(not(any(feature = "jwt-ietf", feature = "jwt-openid")))]
use serde_json::Value as Claims;
use url::Url;
use uuid::Uuid;

type BearerAuthHeader = TypedHeader<Authorization<Bearer>>;

/// Struct representing a JWT, including its [Header]s and `claims`. The `claims` type (`C`) can be
/// customized. If features `jwt-ietf` or `jwt-openid` are enabled, the type will default to
/// the claims for the respective feature. If both features are enabled, the type will default
/// to the claims from `jwt-ietf`. If neither feature is enabled (but `jwt` is enabled), then
/// the default will simply be a [serde_json::Value]. In all cases, the type can be overridden
/// by the consumer.
#[cfg_attr(feature = "open-api", derive(aide::OperationIo))]
// #[derive(Deserialize)]
#[derive(Deserialize, Serialize)]
#[non_exhaustive]
pub struct Jwt<C = Claims> {
    pub header: Header,
    pub claims: C,
}

#[async_trait]
impl<S, C> FromRequestParts<S> for Jwt<C>
where
    S: Clone + Send + Sync + 'static,
    AppContext: FromRef<S>,
    C: for<'de> serde::Deserialize<'de>,
{
    type Rejection = Error;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let context = AppContext::from_ref(state);

        let token = parts
            .extract::<BearerAuthHeader>()
            .await
            .ok()
            .map(|auth_header| auth_header.0.token().to_string());
        let token = if token.is_some() {
            token
        } else {
            let cookies = parts.extract::<CookieJar>().await.ok();
            bearer_token_from_cookies(&context, cookies)
        };

        let token = if let Some(token) = token {
            token
        } else {
            return Err(HttpError::unauthorized().into());
        };

        let token: TokenData<C> = decode_auth_token(
            &token,
            &context.config().auth.jwt.secret,
            &context.config().auth.jwt.claims.audience,
            &context.config().auth.jwt.claims.required_claims,
        )?;
        let token = Jwt {
            header: token.header,
            claims: token.claims,
        };
        Ok(token)
    }
}

fn bearer_token_from_cookies(context: &AppContext, cookies: Option<CookieJar>) -> Option<String> {
    let cookie_name = context
        .config()
        .auth
        .jwt
        .cookie_name
        .clone()
        .unwrap_or_else(|| AUTHORIZATION.to_string());
    cookies
        .as_ref()
        .and_then(|cookies| cookies.get(&cookie_name))
        .map(|cookie| cookie.value())
        .and_then(|token| HeaderValue::from_str(token).ok())
        .and_then(|header_value| {
            <Authorization<Bearer> as axum_extra::headers::Header>::decode(
                &mut [&header_value].into_iter(),
            )
            .ok()
        })
        .map(|auth_header| auth_header.token().to_string())
}

fn decode_auth_token<T1, T2, C>(
    token: &str,
    jwt_secret: &str,
    audience: &[T1],
    required_claims: &[T2],
) -> RoadsterResult<TokenData<C>>
where
    T1: ToString,
    T2: ToString,
    C: for<'de> serde::Deserialize<'de>,
{
    let mut validation = Validation::default();
    validation.set_audience(audience);
    if !required_claims.is_empty() {
        // Todo: Is there a way to reduce the allocations used here?
        let required_claims = validation
            .required_spec_claims
            .iter()
            .map(|claim| claim.to_string())
            .chain(required_claims.iter().map(|claim| claim.to_string()))
            .collect_vec();
        validation.set_required_spec_claims(&required_claims);
    }
    let token_data: TokenData<C> = decode(
        token,
        &DecodingKey::from_secret(jwt_secret.as_ref()),
        &validation,
    )?;
    Ok(token_data)
}

/// The subject of a JWT claim. Technically the IETF spec only specifies that this is a `StringOrURI`
/// type, and the OpenID spec specifies String. However, since this is likely to contain a user ID,
/// we will also try to deserialize directly into a UUID or Integer. Deserialization will fall back
/// to a simple String if the value can not be parsed into a UUID or Integer (or URI).
/// See: <https://www.rfc-editor.org/rfc/rfc7519.html#section-4.1.2>
/// See: <https://openid.net/specs/openid-connect-core-1_0.html#IDToken>
// Intentionally not annotated with `#[non_exhaustive]`
#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(untagged)]
pub enum Subject {
    Uri(Url),
    Uuid(Uuid),
    Int(
        #[serde(
            deserialize_with = "deserialize_from_str",
            serialize_with = "serialize_to_str"
        )]
        u64,
    ),
    String(String),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::snapshot::TestCase;
    use crate::util::serde::Wrapper;
    use axum_extra::extract::cookie::Cookie;
    use insta::assert_debug_snapshot;
    use rstest::{fixture, rstest};
    use serde_json::from_str;
    use std::str::FromStr;
    use url::Url;

    #[fixture]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case::valid_token("Bearer foo")]
    #[case::invalid_token("foo")]
    fn bearer_token_from_cookies(_case: TestCase, #[case] cookie_value: &str) {
        let context = AppContext::test(None, None, None).unwrap();

        let cookies = CookieJar::new().add(Cookie::new(
            AUTHORIZATION.to_string(),
            cookie_value.to_string(),
        ));

        let token = super::bearer_token_from_cookies(&context, Some(cookies));

        assert_debug_snapshot!(token);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_subject_as_uri() {
        let value: Wrapper<Subject> = from_str(r#"{"inner": "https://example.com"}"#).unwrap();
        assert_eq!(
            value.inner,
            Subject::Uri(Url::from_str("https://example.com").unwrap())
        );
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_subject_as_uuid() {
        let uuid = Uuid::new_v4();
        let value: Wrapper<Subject> = from_str(&format!(r#"{{"inner": "{uuid}"}}"#)).unwrap();
        assert_eq!(value.inner, Subject::Uuid(uuid));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_subject_as_int() {
        let num = 100;
        let value: Wrapper<Subject> = from_str(&format!(r#"{{"inner": "{num}"}}"#)).unwrap();
        assert_eq!(value.inner, Subject::Int(num));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn serialize_subject_int_as_string() {
        let num = 100;
        let value = Wrapper {
            inner: Subject::Int(num),
        };
        let s = serde_json::to_string(&value).unwrap();
        assert_eq!(s, format!(r#"{{"inner":"{num}"}}"#));
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_subject_as_string() {
        let value: Wrapper<Subject> = from_str(r#"{"inner": "invalid-uri"}"#).unwrap();
        assert_eq!(value.inner, Subject::String("invalid-uri".to_string()));
    }
}
