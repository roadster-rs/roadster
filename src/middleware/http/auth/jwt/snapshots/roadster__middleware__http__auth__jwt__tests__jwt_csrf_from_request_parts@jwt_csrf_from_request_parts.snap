---
source: src/middleware/http/auth/jwt/mod.rs
expression: jwt
---
{
  "token": {
    "header": {
      "typ": "JWT",
      "alg": "HS256"
    },
    "claims": {
      "iss": null,
      "sub": "[uuid]",
      "exp": 1234,
      "nbf": null,
      "iat": null,
      "jti": null
    }
  },
  "csrf_status": "Safe"
}
